implementation/src/main/java/io/smallrye/mutiny/operators/UniCache.java

```diff
@@ -70,18 +69,50 @@ private synchronized void onCancellation(UniSubscriber<? super I> subscriber) {
         subscribers.remove(subscriber);
     }
 
-    private void replay(UniSubscriber<? super I> subscriber) {
-        synchronized (this) {
-            if (state != COMPLETED) {
-                throw new IllegalStateException(
-                        "Invalid state - expected being in the DONE state, but is in state: " + state);
+    /**
+     * Handling a new subscriber this UniCache is already {@link #COMPLETED}.
+     * Verifies the cached result against the {@link #validator} and
+     * either replays the cached result or renews our own subscription
+     * to restart the cache lifecycle.
+     *
+     * @param subscriber the newly subscribed {@link UniSubscriber}
+     * @return the {@link Runnable} to be executed afterwards
+     */
+    private Runnable handleCompleted(UniSubscriber<? super I> subscriber) {
+        subscribers.add(subscriber);
+
+        // Closure of item and failure for validity checking and replaying the same objects
+        I item = this.item;
+        Throwable failure = this.failure;
+
+        return () -> {
+            boolean isEventValid;
+            synchronized (validatorLock) {
+                // Test whether the cache is still valid
+                // Done within sync-block to avoid multiple subscriptions to the upstream
+                isEventValid = validator.test(item, failure);
             }
-        }
-        if (failure != null) {
-            subscriber.onFailure(failure);
-        } else {
-            subscriber.onItem(item);
-        }
+            if (isEventValid) {
+                // We must first pass a subscription
+                subscriber.onSubscribe(() -> onCancellation(subscriber));
+                // replay the item for that subscriber
+                if (failure != null) {
+                    subscriber.onFailure(failure);
+                } else {
+                    subscriber.onItem(item);
+                }
+            } else {
+                // Items isn't valid any more
+                // Our own subscription has to be renewed
+                synchronized (this) {
+                    state = SUBSCRIBING;
```

Comment:
... and I think here is a "is state already SUBSCRIBING" missing (two cache invalidations raced).
I will build a test for concurrently subscriptions with only invalid items and update it on demand.

Reply:
built in a check if still COMPLETED at time of re-subscription, existing tests working well with that.